framework/include/reporters/ReporterState.h

```diff
@@ -0,0 +1,192 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+#pragma once
+
+#include <iostream>
+
+#include "libmesh/parallel.h"
+#include "libmesh/parallel_object.h"
+
+#include "ReporterName.h"
+#include "RestartableData.h"
+
+/**
+ * A special version of RestartableData to aid in storing Reporter values. This object is
+ * used by the ReporterData object. The objects provides a convenient method to define
+ * Reporter data that has a value as well as some number of old data values as well as
+ * a method for setting the context pointer after construction. Please refer to ReporterData.h
+ * for more information regarding the use of this class.
+ */
+template <typename T>
+class ReporterState : public RestartableData<std::pair<T, std::vector<T>>>
+{
+public:
+  ReporterState(const std::string & name);
+
+  /**
+   * Return a reference to the current value or one of the old values.
+   *
+   * The time_index of 0 returns the current value, 1 returns old, 2 returns older, etc.
+   */
+  T & value(const std::size_t time_index = 0);
+
+  /**
+   * Set the context pointer, see ReporterData for the reason.
+   */
+  void setContext(void * void_ptr);
```

Comment:
This is a WIP. In practice, the ReporterContext objects don't really need to be attached to the the RestartableData since the ReporterData needs to hold them. As with the comment above, the context object have a reference to the restartable data. So, restartable data doesn't need the context or vice versa, something I will play with a bit more.

Reply:
I improved this in the latest commit